x86 Early Boot Fixes

andyross opened the pull request:
Various fixes and cleanups discovered while dealing with early boot on PC hardware.

Commit:
drivers/serial/ns16550: Unbreak 64 bit MMIO addresses

PCI devices on 64 bit systems can be mapped anywhere, not just in the
lower 4G of memory.  Remove pointer size assumptions.

Also this removes the use of a struct uart_device_config to store the
(runtime) BAR address.  That struct has other stuff in it, and the
only thing we need is the single MMIO address.  It's also REALLY
confusing to have two "devconf" fields in the device storing values of
the same struct type, some fields of which are used from one of them
but some from the other!

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
drivers/pcie: Fix BAR address size limitation

The PCI API was originally limited to 32 bit addresses.  Even though
it had code to skip over the high word in 64 bit BAR entries, it
refused to use it and returned a 32 bit value.  Some devices in the
wild have default mappings from the firmware for devices above 4G.

Also remove the "iobar" API.  It's dead code, we don't call it and we
don't test it.  IO space BAR entries are a legacy feature from way,
way back in PCI history (I genuinely have never heard of a real device
that uses them!).  And there's no difference in format between one of
these and a 32 bit "memory" BAR anyway, someone who actually had this
requirement could just use the existing API without worry.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
arch/x86: Map 512G by default when booting x86_64

The default page table (the architecturally required one used for
entrance to long mode, before the OS page tables get assembled) was
mapping the first 4G of memory.

Extend this to 512G by fully populating the second level page table.
We have devices now (up_squared) which have real RAM mapped above 4G.
There's really no good reason not to do this, the page is present
always anyway.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
sys/sys_io.h: Memory mapped I/O APIs must be pointer-sized, not 32 bit!

The mm_reg_t type used to represent a MMIO address (e.g. in the
sys_write*() APIs) was defined to a uint32_t, which is obviously wrong
on 64 bit systems where devices can be mapped anywhere.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
arch/x86: early_serial cleanup

Various cleanups to the x86 early serial driver, mostly with the goal
of simplifying its deployment during board bringup (which is really
the only reason it exists in the first place):

+ Configure it =y by default.  While there are surely constrained
  environments that will want to disable it, this is a TINY driver,
  and it serves a very important role for niche tasks.  It should be
  built always to make sure it works everywhere.

+ Decouple from devicetree as much as possible.  This code HAS to work
  during board bringup, often with configurations cribbed from other
  machines, before proper configuration gets written.  Experimentally,
  devicetree errors tend to be easy to make, and without a working
  console impossible to diagnose.  Specify the device via integer
  constants in soc.h (in the case of IOPORT access, we already had
  such a symbol) so that the path from what the developer intends to
  what the code executes is as short and obvious as possible.
  Unfortunately I'm not allowed to remove devicetree entirely here,
  but at least a developer adding a new platform will be able to
  override it in an obvious way instead of banging blindly on the
  other side of a DTS compiler.

+ Don't try to probe the PCI device by ID to "verify".  While this
  sounds like a good idea, in practice it's just an extra thing to get
  wrong.  If we bail on our early console because someone (yes, that's
  me) got the bus/device/function right but typoed the VID/DID
  numbers, we're doing no one any favors.

+ Remove the word-sized-I/O feature.  This is a x86 driver for a PCI
  device.  No known PC hardware requires that UART register access be
  done in dword units (in fact doing so would be a violation of the
  PCI specifciation as I understand it).  It looks to have been cut
  and pasted from the ns16550 driver, remove.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
arch/x86/early_serial: General code cleanup

This patch is almost entirely aesthetics, designed to isolate the
variant configurations to a simple macro API (just IN/OUT), reduce
complexity derived from code pasted out of the larger ns16550 driver,
and keep the complexity out of the (very simple!) core code.  Useful
when hacking on the driver in contexts where it isn't working yet.

The sole behavioral change here is that I've removed the runtime
printk hook installation in favor of defining an
arch_printk_char_out() function which overrides the weak-linked
default (that is, we don't need to install a hook, we can be the
default hook at startup).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
arch/x86: Add a spurious interrupt handler to x86_64

Right now x86_64 doesn't install handlers for vectors that aren't
populated by Zephyr code.  Add a tiny spurious interrupt handler that
logs the error and triggers a fatal error, like other platforms do.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andrewboie commented on arch/x86/core/early_serial.c:
```diff
@@ -1,102 +1,67 @@
 }
 
-static int console_out(int c)
+int arch_printk_char_out(int c)
```
If we're adding a new arch interface, this must be prototyped and documented in the arch interface headers.

I might be missing something obvious (was there a patch that got dropped?) but nothing in this PR references this function so I don't see how this works. I prefer that we left this and the invocation of `__printk_hook_install()` as-is.

andrewboie replied:
Heyyyy I'm just dumb...arch_printk_char_out() infrastructure is already in place. disregard! this is better.

andrewboie commented on arch/x86/core/early_serial.c:
```diff
@@ -1,102 +1,67 @@
+#define OUT(reg, val) sys_write8(val, X86_SOC_EARLY_SERIAL_MMIO8_ADDR + reg)
 #endif
 
+#define REG_THR  0x00  /* Transmitter holding reg. */
```
I wonder if it makes sense to introduce a common ns16550.h with these definitions under drivers/serial and have both this file and the real 16550 driver use it.

andrewboie commented on arch/x86/core/prep_c.c:
```diff
@@ -44,6 +49,8 @@ FUNC_NORETURN void z_x86_prep_c(void *arg)
 			    true);
 #endif
 
+	x86_irq_init();
```
I find it more aesthetically pleasing to ifdef this under CONFIG_X86_64 rather than having to define an empty weak function earlier in this file so that the 32-bit build doesn't break, that weak function confused me for a bit. I don't care that much though.

tyrantlizard77 reviewed (changes requested):
I'm seeing something weird when I test this.

If I build and run a test locally (any test, doesn't seem to matter) I get a hung console:
```
SeaBIOS (version rel-1.12.1-0-ga5cab58-dirty-20200516_152813-e7ee184eb100-zephyr
)
Booting from ROM..
```

But if I run sanitycheck, sanitycheck passes. I think this is a bug in sanitycheck. And a really bad one, because the handler.log in the sanity-out directory for every single test just has the SEABios message, but no test output, yet it is reporting success(!!).

stephanosio commented:
> I'm seeing something weird when I test this.
> 
> If I build and run a test locally (any test, doesn't seem to matter) I get a hung console:
> 
> ```
> SeaBIOS (version rel-1.12.1-0-ga5cab58-dirty-20200516_152813-e7ee184eb100-zephyr
> )
> Booting from ROM..
> ```

Reminds me of this, might be related:
https://github.com/zephyrproject-rtos/sdk-ng/issues/168
https://github.com/zephyrproject-rtos/zephyr/pull/21733

andrewboie commented on arch/x86/core/early_serial.c:
```diff
@@ -105,30 +70,20 @@ static int console_out(int c)
-		FCR_FIFO | FCR_FIFO_1 | FCR_RCVRCLR | FCR_XMITCLR);
-
-	/* Will be replaced later when a real serial driver comes up */
-	__printk_hook_install(console_out);
```
disregard!

tyrantlizard77 reviewed (approved):
Never mind, `arch_printk_char_out()` is already a thing. This looks fine other than a few nits I pointed out and need to find out why boot is getting stuck.

andrewboie commented:
Sanitycheck issue addressed by https://github.com/zephyrproject-rtos/zephyr/pull/26223, I used this patch to help test it.

andyross commented:
Yeah, that disassembly is showing a port of "5", which is the register offset.  I clearly broke the ioport generator, will fix.  This was sort of a rush job to get it into a PR-worthy state and almost all my testing was on MMIO hardware.

andyross commented:
Rebase, address comments, fix one goof that had broken irq_connect_dynamic() (which apparently we only enable for one test?)

andyross commented:
Close/open again to kick CI.  Seems like the buildkit error is stale?  It's reflecting a failure the previous push corrected.

tbursztyka commented on soc/x86/apollo_lake/soc.h:
```diff
@@ -27,4 +27,9 @@
 
+#include <drivers/pcie/pcie.h>
+
+#define X86_SOC_EARLY_SERIAL_PCIDEV PCIE_BDF(0, 0x18, 0)
```
Shouldn't these info provided via dts?

andyross replied:
Getting early configuration **out** of devicetree was an explicit design goal here.  A value in DTS needs to go through multiple levels of compilers before it gets used in code, and without a console there's no way to debug that.

This is the console driver you use to debug your DTS hacking, basically.   If we wanted design purity, it wouldn't exist at all. :)

nashif replied:
not sure I follow, the values in DTS are available early during build time and before we actually start building any code

> A value in DTS needs to go through multiple levels of compilers before it gets used in code

This is not correct AFAIK, DTS data is available just like Kconfig data, there is nothing that is runtime here, all data is available in header files before we start building. Maybe I am missing something here.

nashif replied:
We should get this from DTS:

```DT_REG_ADDR(DT_CHOSEN(zephyr_console))```

andyross replied:
It's not about technical capability, it's about human reasoning capability.  My ability to know that this code is doing the right thing is very seriously hampered by having to follow it all the way through  a ton of external tooling.  And this is experimentally true, given that I personally made multiple mistakes trying to get the DTS values correct (c.f. the other change in this patch that unbreaks the driver when you write "4d4b" instead of "4b4d").

Again, this is the driver you use when all the drivers don't work.  Coupling it to other driver subsystems defeats the purpose.  If you really want all systems using the same set of frameworks, we have a console driver that does that already.  And when it didn't work, I needed to use this one to find out why.

andrewboie replied:
Could we split the difference on this DTS concern? That is to day, default to a set of #defines for the addr/size or bdf/id based on what DT_REG_ADDR/DT_REG_SIZE return from DTS for "zephyr_console", but provide a blindingly obvious way to override it if it doesn't seem to be working for some reason?

In other words, in early_serial.c something along the lines of:

```
#ifndef X86_SOC_EARLY_SERIAL_PCIDEV
#define X86_SOC_EARLY_SERIAL_PCIDEV DT_REG_ADDR(zephyr_console)
```

And if it doesn't work out of the box for whatever reason set the define in soc.h to un-block debugging?

> Experimentally, devicetree errors tend to be easy to make, and without a working console impossible to diagnose

This is fair that working with DTS can be annoying. Are there changes we could make to make debugging DTS issues easier? Everything is known at build time, we could for example produce a human readable report in the build directory so that if you want to know what the DTS data is for "zephyr_console" you can just look it up.

It is definitely the case that right now, generated files like devicetree_unfixed.h are incomprehensible.

Having said all this, I won't lose sleep if we merge as-is as long as we have an understanding that doing this in this driver is a very special case, this PR retains my +1.

nashif replied:
> Could we split the difference on this DTS concern? That is to day, default to a set of #defines for the addr/size or bdf/id based on what DT_REG_ADDR/DT_REG_SIZE return from DTS for "zephyr_console", but provide a blindingly obvious way to override it if it doesn't seem to be working for some reason?
> 
> In other words, in early_serial.c something along the lines of:
> 
> ```
> #ifndef X86_SOC_EARLY_SERIAL_PCIDEV
> #define X86_SOC_EARLY_SERIAL_PCIDEV DT_REG_ADDR(zephyr_console)
> ```
> 
> And if it doesn't work out of the box for whatever reason set the define in soc.h to un-block debugging?


yes, that works.

andyross commented:
Just for clarity:

> We should get this from DTS:

Under absolutely no circumstances.  I'll go all the way to the mattresses on this.  If you want all x86 console drivers to use DTS then submit a patch to remove this one.

nashif commented:
> Under absolutely no circumstances. I'll go all the way to the mattresses on this. If you want all x86 console drivers to use DTS then submit a patch to remove this one.

the driver was already using DTS and you are changing it to use a value that is hardcoded in the drive, what is the rationale behind all of this? 

```
diff --git a/arch/x86/core/early_serial.c b/arch/x86/core/early_serial.c
index c4d9cc4f29..58dcebef5b 100644
--- a/arch/x86/core/early_serial.c
+++ b/arch/x86/core/early_serial.c
@@ -4,19 +4,19 @@
  * SPDX-License-Identifier: Apache-2.0
  */

-#include <devicetree.h>
 #include <kernel.h>
 #include <sys/util.h>
+#include <drivers/pcie/pcie.h>

 #include <soc.h>

-#if DT_PROP(DT_CHOSEN(zephyr_console), pcie)
-BUILD_ASSERT(IS_ENABLED(CONFIG_PCIE), "NS16550(s) in DT need CONFIG_PCIE");
-#define UART_NS16550_PCIE_ENABLED
-#include <drivers/pcie/pcie.h>
-
-#define UART_PCIE_BDF  (DT_REG_ADDR(DT_CHOSEN(zephyr_console)))
-#define UART_PCIE_ID   (DT_REG_SIZE(DT_CHOSEN(zephyr_console)))
+#if defined(X86_SOC_EARLY_SERIAL_PCIDEV)
+#define UART_PCIE_BDF X86_SOC_EARLY_SERIAL_PCIDEV
+#define UART_NS16550_PCIE_ENABLED 1
+#elif defined(UART_NS16550_ACCESS_IOPORT)
+#undef UART_NS16550_PCIE_ENABLED
+#else
+#error "Incomplete x86 SoC early serial config"
 #endif
```

andyross commented:
@nashif Please don't start a fight on this.  The reasoning is called out explicitly in the commit message, and I explained it above again.

I won't make that change to this PR, it's just technically wrong.  This driver doesn't have value if it's coupled to devicetree.  If that's your position and you won't budge, then just remove it.  We have another console driver.  We only need this one if we need to bring up more boards that defeat analysis.

jhedberg commented:
Could we try to get this moving forward? @andyross there seems to be conflicts with `arch/x86/core/intel64/locore.S` - possibly due to the zefi PR getting merged?

andyross commented:
Oh right, I actually uncollided these to submit them separately and one of the PRs was always going to need a rebase.  Will fix ASAP.  Note that there's a rebased variant of these patches in my ehl-wip branch if you need it sooner.

Commit:
arch/x86: zefi must disable HPET before OS handoff

The firmware on existing devices uses HPET timer zero for its own
purposes, and leaves it alive with interrupts enabled.  The Zephyr
driver now knows how to recover from this state with fuller
initialization, but that's not enough to fix the inherent race:

The timer can fire BEFORE the driver initialization happens (and does,
with certain versions of the EFI shell), thus flagging an interrupt to
what Zephyr sees as a garbage vector.  The OS can't fix this on its
own, the EFI bootloader (which is running with interrupts enabled as
part of the EFI environment) has to do it.  Here we can know that our
setting got there in time and didn't result in a stale interrupt flag
in the APIC waiting to blow up when interrupts get enabled.

Note: this is really just a workaround.  It assumes the hardware has
an HPET with a standard address.  Ideally we'd be able to build zefi
using Zephyr kconfig and devicetree values and predicate the HPET
reset on the correct configuraiton.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
New version includes an extra patch to zefi to workaround a fundamental race with HPET.  Please re-review as needed. Existing comments should be addressed.

Commit:
arch/x86: Make sure PCI mmio is initialized before page tables

The page table initialization needs a populated PCI MMIO
configuration, and that is lazy-evaluated.  We aren't guaranteed that
a driver already hit that path, so be sure to call it explicitly.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>
